### PR TITLE
Tree view and semantic information with `Whiff` views

### PR DESCRIPTION
**Pull request by mhhf**
halp

### CODE REVIEW

**Comment by MrChico** [comment]
Exposes the ability to print the execution tree built from `hevm symbolic` via the `--show-tree` flag.
Example usage:
```
./src/hevm/dist-newstyle/build/x86_64-osx/ghc-8.6.5/hevm-0.41.0/x/hevm/build/hevm/hevm symbolic --code $(cat combined.json | jq  '.contracts."token.sol:Token"."bin-runtime"' | tr -d '"') --json-file combined.json --solver cvc4 --show-tree
checking postcondition...
Q.E.D.
Explored: 15 branches without assertion violations
├ 0    isZero(isZero(<CallValue>))
│         Revert
└ 1    isZero(isZero(<CallValue>))
  ├ 0    isZero((<Calldatasize> < 0x4))
  │ ├ 0    isZero((0x70a08231 == <symbolic>))
  │ │ ├ 0    isZero((0xa9059cbb == <symbolic>))
  │ │ │         Revert
  │ │ └ 1    isZero((0xa9059cbb == <symbolic>))
  │ │   ├ 0    isZero(isZero(((<Calldatasize> + -(0x4)) < 0x40)))
  │ │   │         Revert
  │ │   └ 1    isZero(isZero(((<Calldatasize> + -(0x4)) < 0x40)))
  │ │     ├ 0    isZero(isZero((<symbolic> > <symbolic>)))
  │ │     │         Revert
  │ │     └ 1    isZero(isZero((<symbolic> > <symbolic>)))
  │ │               Stop
  │ └ 1    isZero((0x70a08231 == <symbolic>))
  │   ├ 0    isZero(isZero(((<Calldatasize> + -(0x4)) < 0x20)))
  │   │         Revert
  │   └ 1    isZero(isZero(((<Calldatasize> + -(0x4)) < 0x20)))
  │             Return: 32 symbolic bytes
  └ 1    isZero((<Calldatasize> < 0x4))
            Revert
```

We maintain an AST of stack elements using [Whiff](https://github.com/dapphub/dapptools/pull/533/files#diff-89cea9f0ead2c601fb68ea4ae2d4e2baf08d983e473c7b106282cf43cb16f267R42), which is also pretty printed if the term is symbolic. In other words, where we currently see only `<symbolic>` on the stack, we now see expressions like `isZero((<symbolic> > <symbolic>))`, (here corresponding stemming from `balanceOf[msg.sender] <= value`). 

Later PRs will add `Whiff` to buffers as well, enabling us to print variable names and storage locations taken from the AST. Later this tree can also be pretty printed as an `act`

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 1]
```diff
@@ -774,24 +788,24 @@ symvmFromCommand cmd = do
                                    )
 
   caller' <- maybe (SAddr <$> freshVar_) (return . litAddr) (caller cmd)
-  ts <- maybe (sw256 <$> freshVar_) (return . w256lit) (timestamp cmd)
-  callvalue' <- maybe (sw256 <$> freshVar_) (return . w256lit) (value cmd)
+  ts <- maybe (S (Var "Timestamp") <$> freshVar_) (return . w256lit) (timestamp cmd)
+  callvalue' <- maybe ((S (Var "CallValue")) <$> freshVar_) (return . w256lit) (value cmd)
   (calldata', cdlen, pathCond) <- case (calldata cmd, sig cmd) of
     -- fully abstract calldata (up to 1024 bytes)
```
Comment (xwvvvvwx) [inline src/hevm/hevm-cli/hevm-cli.hs:794]: This comment is outdated right? I think it should be:

```suggestion
    -- fully abstract calldata (up to 256 bytes)
```

File: src/hevm/src/EVM.hs [hunk 2]
```diff
@@ -624,18 +624,18 @@ exec1 = do
         0x09 -> stackOp3 (const g_mid) (\(x, y, z) -> mulmod x y z)
 
         -- op: LT
-        0x10 -> stackOp2 (const g_verylow) $ \(x, y) -> ite (x .< y) 1 0
+        0x10 -> stackOp2 (const g_verylow) $ \(x, y) -> iteWhiff "<" (x .< y) x y
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:627]: why change from `1 0` to `x y` here?

File: src/hevm/src/EVM.hs [hunk 3]
```diff
@@ -675,13 +675,15 @@ exec1 = do
...
+                                     -- TODO - maybe here Dull needs to be replaced with something smarrrtttt
 
                                      SymbolicBuffer bs -> do
                                                    let hash' = symkeccak' bs
                                                        previousUsed = view (env . keccakUsed) vm
                                                    env . keccakUsed <>= [(bs, hash')]
-                                                   pathConditions <>= fmap (\(preimage, image) ->
-                                                                               image .== hash' .=> preimage .== bs)
+                                                   constraints <>= fmap (\(preimage, image) ->
+                                                                               (image .== hash' .=> preimage .== bs, Dull))
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:686]: can't we use `FromKeccak preimage` instead of `Dull` here?

File: src/hevm/src/EVM.hs [hunk 4]
```diff
@@ -1535,20 +1537,20 @@ askSMT codeloc condition continue = do
...
+     Nothing -> do pathconds <- use constraints
                    assign result . Just . VMFailure . Query $ PleaseAskSMT
-                     condition pathconds choosePath
+                     condition (fst <$> pathconds) choosePath
 
    where -- Only one path is possible
          choosePath :: BranchCondition -> EVM ()
          choosePath (Case v) = do assign result Nothing
-                                  pushTo pathConditions (if v then condition else sNot condition)
+                                  pushTo constraints (if v then (condition, UnOp "not" whiff) else (sNot condition, whiff))
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:1547]: should the `not` be in the else branch here?

File: src/hevm/src/EVM/SymExec.hs [hunk 5]
```diff
@@ -133,6 +137,43 @@ loadSymVM x initStore model addr callvalue' calldata' =
     }) & set (env . contracts . at (createAddress ethrunAddress 1))
              (Just (contractWithStore x initStore))
 
+-- todo branchinfo needs to contain two vm's,
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/SymExec.hs:140]: why is this needed?

File: src/hevm/src/EVM/SymExec.hs [hunk 6]
```diff
@@ -225,33 +266,58 @@ pruneDeadPaths =
...
+    return $ Just (Node b consistentChildren)
+    
+  
+leaves :: Tree BranchInfo -> [VM]
+leaves (Node x []) = [_vm x]
+leaves (Node _ xs) = concatMap leaves xs
+
 -- | Symbolically execute the VM and check all endstates against the postcondition, if available.
 -- Returns `Right VM` if the postcondition can be violated, where `VM` is a prestate counterexample,
 -- or `Left (VM, [VM])`, a pair of `prestate` and post vm states.
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/SymExec.hs:296]: outdated comment.

File: src/hevm/src/EVM/SymExec.hs [hunk 7]
```diff
@@ -225,33 +266,58 @@ pruneDeadPaths =
     Just (VMFailure DeadPath) -> False
     _ -> True
 
+consistentPath :: VM -> Query (Maybe VM)
+consistentPath vm = do
+  resetAssertions
+  constrain $ sAnd $ fst <$> view constraints vm
+  checkSat >>= \case
+    Sat -> return $ Just vm
+    _   -> return $ Nothing
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/SymExec.hs:275]: do we maybe want to distinguish between unsat results and timeouts here?

File: src/hevm/src/EVM/Format.hs [hunk 8]
```diff
@@ -319,3 +322,69 @@ contractNamePart x = Text.split (== ':') x !! 1
...
+  then "Stop"
+  else "Return: " <> show (ByteStringS msg)
+prettyvmresult (EVM.VMSuccess (SymbolicBuffer msg)) =
+  "Return: " <> show (length msg) <> " symbolic bytes"
+
+-- TODO
+-- display in an 'act' was - propagate iff and if's, prune failed paths
+-- display cases instead of branches - collapse cases
+-- display storage changes (M&D)
+-- display condition correctly
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/Format.hs:342]: what is incorrect atm?

**Review by xwvvvvwx — COMMENTED** [review]
a few little nits and questions, but otherwise looks awesome!!

File: src/hevm/src/EVM.hs (hunk 2 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs:627]: `iteWhiff` returns 0 or 1, but with the whiff built out of x and y

File: src/hevm/src/EVM.hs (hunk 3 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs:686]: kind of.... It's more like a injectivity constraint on keccak than a fromKeccak

File: src/hevm/src/EVM/SymExec.hs [hunk 9]
```diff
@@ -133,6 +137,43 @@ loadSymVM x initStore model addr callvalue' calldata' =
     }) & set (env . contracts . at (createAddress ethrunAddress 1))
              (Just (contractWithStore x initStore))
 
+-- todo branchinfo needs to contain two vm's,
+-- the start vm, and the vm before the branching point/ at the (endstate/leaf)
```
Comment (MrChico) [inline src/hevm/src/EVM/SymExec.hs:141]:
```suggestion
```

File: src/hevm/src/EVM/SymExec.hs (hunk 7 above)
Comment (MrChico) [reply src/hevm/src/EVM/SymExec.hs:275]: yes, good idea!

File: src/hevm/src/EVM.hs (hunk 4 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs:1547]: yeah, I think so

**Comment by MrChico** [comment]
thx for review @xwvvvvwx your comments should be addressed

**Review by xwvvvvwx — APPROVED** [review]
lgtm! :sparkles:
